### PR TITLE
Add exponential backoff for erroring jobs

### PR DESCRIPTION
**Pull request by sagelywizard**
This PR adds an exponential backoff for rescheduling erroring jobs. It does this by counting the number of consecutive times which a job has errored in its recorded history (up to a maximum) and using that to calculate the amount of time to wait before rescheduling the job.

This PR also adds a config parameter which is important for exponential backoff: `max_history`. This sets the maximum number of `started`, `stopped`, and `crashed` events which the job will track. Setting a large history will allow the job to be suspended for longer (up to a configurable maximum) but will cause higher memory usage.

### CODE REVIEW

File: src/couch_replicator_scheduler.erl [hunk 1]
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
...
+    ets:select(?MODULE, MatchSpec).
 
 
 -spec pending_job_count() -> non_neg_integer().
 pending_job_count() ->
-    MatchSpec = [{#job{pid='$1', _='_'}, [{'not', {'is_pid', '$1'}}], [true]}],
+    Now = unix_now(),
+    % Select jobs where the pid is undefined and the exponential backoff
+    % threshold has been crossed.
+    MatchSpec = [{#job{pid = '$1',errored_at = '$2',errors = '$3', _='_'},
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Should we use fun2ms here, for clarity? It was a bit tricky deciphering the meaning of this expression

File: src/couch_replicator_scheduler.erl [hunk 2]
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
...
+        [{'not',{is_pid,'$1'}},
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Maybe in the head just say `#job{pid = undefined}` to save an extra guard. Should probably also update typespec of job records' pid field to be `pid() | undefined`

File: src/couch_replicator_scheduler.erl [hunk 3]
```diff
@@ -259,25 +267,57 @@ remove_job_int(#job{} = Job) ->
...
 
+-spec job_count() -> non_neg_integer().
+job_count() ->
+    ets:info(?MODULE, size).
+
+
 -spec running_job_count() -> non_neg_integer().
 running_job_count() ->
-    ets:info(?MODULE, size) - pending_job_count().
+    MatchSpec = [{#job{pid = '$1', _='_'},
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Would fun2ms or qlc here be too hard on performance?

File: src/couch_replicator_scheduler.erl [hunk 4]
```diff
@@ -354,3 +394,9 @@ update_history(Job, Type, When) ->
     History0 = [{Type, When} | Job#job.history],
     History1 = lists:sublist(History0, ?MAX_HISTORY),
     Job#job{history = History1}.
+
+
+-spec unix_now() -> float().
+unix_now() ->
+    {Meg, Sec, Mic} = os:timestamp(),
+    Meg*1000000+Sec+(Mic/1000000).
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Minor nit: spacing between operators and operands. For consistency.

File: src/couch_replicator_scheduler.erl [hunk 5]
```diff
@@ -30,23 +30,25 @@
...
 -type history() :: [Events :: event()].
 
 %% definitions
 -define(MAX_HISTORY, 20).
--define(MINIMUM_CRASH_INTERVAL, 60 * 1000000).
 
 -define(DEFAULT_MAX_JOBS, 100).
 -define(DEFAULT_MAX_CHURN, 20).
 -define(DEFAULT_SCHEDULER_INTERVAL, 60000).
+-define(ERROR_BACKOFF_MILLIS, 30000).
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Saw below we calculate back-off interval in seconds and then translate to milliseconds to compare with this constant.  Maybe just use seconds here? ... Or would that clash visually with `DEFAULT_SCHEDULER_INTERVAL` which is in milliseconds...

File: src/couch_replicator_scheduler.erl [hunk 6]
```diff
@@ -30,23 +30,25 @@
...
 -record(state, {interval, timer, max_jobs, max_churn}).
 -record(job, {
           id :: job_id(),
           rep :: #rep{},
           pid :: pid(),
-          history :: history()}).
+          history :: history(),
+          errors :: integer(),
+          errored_at :: integer()}).
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Is this an integer or a float? Initialized below to 0.0 but not sure if it we still mean to be float.

**Comment by nickva** [comment]
Should we squash the commits in the pr?  Most are to the same file or same code blocks. Fixes during review could be separate commits if we want to track that.

**Comment by nickva** [comment]
For exponential back-off I think `uniform_random(2^ConsecutiveErrorCount)` might be better. TCP uses something like that. 

The difference being it avoids global synchronization between jobs. Say jobs replicate from one source, it fails, then all jobs will start retrying at the same schedule.

**Comment by nickva** [comment]
To avoid the complicated multi-line matchspecs, was wondering if perhaps it is better to use a `next_try_at` field in the job record?

When job stops, depending how it stops, we schedule it then when to run next by setting `next_try_at = uniform_random(2^ConsecutiveErrorCount * ?ERROR_BACKOFF)`. 

Then we don't have to to compute all that stuff during match spec and it just compares next_try_at with Now.

File: src/couch_replicator_scheduler.erl (hunk 6 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Should've been `float`.

File: src/couch_replicator_scheduler.erl (hunk 2 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Good thought.

**Comment by sagelywizard** [comment]
For the `try_next_at` field idea, I had thought of that but it makes it difficult to change the delay interval configuration dynamically. I dunno if that's a big deal, but it's why I went for the matchspec approach.

**Comment by sagelywizard** [comment]
Yeah, I'll definitely squash this PR down once I get a "+1 pending squash".

**Comment by sagelywizard** [comment]
Good point about the random uniform thing. Though do we need it since we have the max churn stuff? It'd require a `try_next_at` field and I'm not sure we want to do that. @rnewson @nickva Thoughts?

File: src/couch_replicator_scheduler.erl [hunk 7]
```diff
@@ -207,14 +207,17 @@ not_recently_crashed(#job{} = Job) ->
     case crash_history(Job) of
         [] ->
             true;
-        [{crashed, When} | _] ->
+        [{{crashed, _Reason}, When} | _] = Crashes ->
+            BackoffExp = erlang:min(1 bsl length(Crashes), ?MAX_BACKOFF_EXPONENT),
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: `?MAX_BACKOFF_EXPONENT` is not the exponent though, `length(Crashes)` is the exponent.

Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Hrm, `?MAX_BACKOFF_EXPONENT` may be a bit of a misnomer, but the code looks correct to me. Could be `1 bsl erlang:min(length(Crashes), ?MAX_BACKOFF_EXPONENT)`. I'll make that change.

File: src/couch_replicator_scheduler.erl [hunk 8]
```diff
@@ -266,57 +260,24 @@ remove_job_int(#job{} = Job) ->
...
-    % Select jobs where the pid is undefined and the exponential backoff
-    % threshold has been crossed.
-    MatchSpec = [{#job{pid = '$1',errored_at = '$2',errors = '$3', _='_'},
-        [{'not',{is_pid,'$1'}},
-        {'<',{'-',{'*',{'bsl',1,'$3'},?ERROR_BACKOFF_MILLIS},
-        {'*',{'-',{const,Now},'$2'},1000}},
-        0}],
-        [true]}
-    ],
+    MatchSpec = [{#job{pid='$1', _='_'}, [{'not', {'is_pid', '$1'}}], [true]}],
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Should this  be `#job{pid=undefined, ...}` ? 

Going by idea that pending jobs are the ones which are not running but a ready to run, so they don't have a pid yet.

Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Changed

File: src/couch_replicator_scheduler.erl [hunk 9]
```diff
@@ -266,57 +260,24 @@ remove_job_int(#job{} = Job) ->
...
+    ets:info(?MODULE, size) - pending_job_count().
 
 -spec running_jobs() -> [#job{}].
 running_jobs() ->
-    MatchSpec = [{#job{pid = '$1', _='_'},
-        [{is_pid,'$1'}],
-        ['$_']}
-    ],
-    ets:select(?MODULE, MatchSpec).
+    ets:tab2list(?MODULE) -- pending_jobs().
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Superficially looking at it, this could be costly. First we'd make a full copy of all the jobs in tab2list. Then a copy in pending_jobs(). Then another copy is the resulting list. L1--L2 complexity is about O(length(L1)*length(L2)) as well.

Since we are returning whole objects perhaps `ets:match_object(Tab, Pattern)` would work here.

Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Yeah, it is rather costly. I don't think you could use `ets:match_object` since we want to filter by type. The matchspec/select which I reverted is less costly, so I could un-revert that.

**Comment by sagelywizard** [comment]
That was a rebase onto 63012-scheduler

File: src/couch_replicator_scheduler.erl [hunk 10]
```diff
@@ -225,17 +232,22 @@ oldest_job_first(#job{} = A, #job{} = B) ->
...
+            couch_log:notice("delay for ~p is now ~p s", [Job#job.id, BackoffInterval div 1000000]),
             timer:now_diff(os:timestamp(), When)
-                >= ?MINIMUM_CRASH_INTERVAL
+                >= BackoffInterval;
+        [] ->
+            true
     end.
 
 
 crash_history(#job{} = Job) ->
```
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: I think we want `JobHistoryEvents :: list()` here as arg. Or maybe it is cleaner to push EventsBeforeStop processing to `crash_history/ 1` to make not_recently_crashed a bit smaller visually.

**Comment by nickva** [comment]
+1

File: src/couch_replicator_scheduler.erl [hunk 11]
```diff
@@ -148,8 +149,14 @@ handle_info({'DOWN', _Ref, process, Pid, Reason}, State) ->
         {ok, #job{}=Job0} ->
             couch_log:notice("~p: Job ~p died with reason: ~p",
                              [?MODULE, Job0#job.id, Reason]),
-            Job1 = update_history(Job0#job{pid = undefined}, crashed, os:timestamp()),
-            true = ets:insert(?MODULE, Job1),
+            Job1 = Job0#job{pid = undefined},
+            Job2 = case Reason of
+                normal ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: given the previous function clause, how could Reason be normal here?

File: src/couch_replicator_scheduler.erl [hunk 12]
```diff
@@ -30,13 +30,14 @@
 -export([handle_config_change/5, handle_config_terminate/3]).
 
 %% types
--type event_type() :: started | stopped | crashed.
+-type event_type() :: started | stopped | {crashed, any()}.
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:33]: where is the Reason term used? since it'll often be large, and we'll have lots of jobs in memory I think we might not want to store it.

File: src/couch_replicator_scheduler.erl [hunk 13]
```diff
@@ -284,13 +296,12 @@ running_job_count() ->
 
 -spec running_jobs() -> [#job{}].
 running_jobs() ->
-    ets:tab2list(?MODULE) -- pending_jobs().
+    ets:select(?MODULE, [{#job{pid = '$1', _='_'}, [{is_pid, '$1'}], ['$_']}]).
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:316]: nice. can we pull this into a separate commit, though? (since it's not about adding exponential backoff)

File: src/couch_replicator_scheduler.erl [hunk 14]
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
 
 not_recently_crashed(#job{} = Job) ->
     case crash_history(Job) of
-        [] ->
-            true;
-        [{crashed, When} | _] ->
+        [{{crashed, _Reason}, When} | _] = Crashes ->
+            BackoffExp = erlang:min(length(Crashes), ?MAX_BACKOFF_EXPONENT),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: off by one here. the first time we back off will be for _two_ backoff intervals not one. suggest length(Crashes) - 1.

File: src/couch_replicator_scheduler.erl [hunk 15]
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
...
+            BackoffInterval = (1 bsl BackoffExp) * ?BACKOFF_INTERVAL_MICROS,
+            couch_log:notice("delay for ~p is now ~p s", [Job#job.id, BackoffInterval div 1000000]),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: remove log line, this'll just be noise. Perhaps move the logic here to a separate side-effect free function instead, so we can write some eunit tests to prove the progression of timeouts (and their cap).

File: src/couch_replicator_scheduler.erl (hunk 10 above)
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: I think crash_history should return the crash history, we've mingled in a separate concern here.

File: src/couch_replicator_scheduler.erl [hunk 16]
```diff
@@ -226,16 +233,21 @@ oldest_job_first(#job{} = A, #job{} = B) ->
 
 not_recently_crashed(#job{} = Job) ->
     case crash_history(Job) of
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: if the most recent history item is a `stopped`, we could return true here. That is, we can define that a job that last ran successfully has not recently crashed.

File: src/couch_replicator_scheduler.erl (hunk 14 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Or make backoff =  random:uniform(interval)? The expected back off will be one interval.

File: src/couch_replicator_scheduler.erl (hunk 10 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Agree. Was thinking with my 10pm brain there trying to get tests to pass. If crash history is not called from anywhere else, the fix could rename the function, as well...

File: src/couch_replicator_scheduler.erl (hunk 11 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Whoops, I'll fix that.

File: src/couch_replicator_scheduler.erl (hunk 13 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl:316]: Sure

File: src/couch_replicator_scheduler.erl (hunk 15 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: K

File: src/couch_replicator_scheduler.erl (hunk 12 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl:33]: For 1) introspection and 2) the ability to make scheduling decisions based on specific errors. Thoughts?

File: src/couch_replicator_scheduler.erl (hunk 14 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: I think random would make sense if the backoff exponent was persisted in memory. If the backoff was randomly generation for each filtration of crashed jobs, jobs will be biased toward low exponents.

File: src/couch_replicator_scheduler.erl (hunk 16 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Sure.

File: src/couch_replicator_scheduler.erl (hunk 12 above)
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:33]: ok

File: src/couch_replicator_scheduler.erl [hunk 17]
```diff
@@ -129,28 +142,23 @@ handle_cast(_, State) ->
...
     {noreply, State#state{timer = Timer}};
 
-handle_info({'DOWN', _Ref, process, Pid, normal}, State) ->
-    case job_by_pid(Pid) of
-        {ok, #job{}=Job} ->
-            remove_job_int(Job);
-        _Else ->
-            ok
-    end,
+handle_info({'DOWN', _Ref, process, _Pid, normal}, State) ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: why remove the job cleanup when a replication completes?

File: src/couch_replicator_scheduler.erl [hunk 18]
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
...
 %% private functions
 
-start_jobs(Count) ->
+start_jobs(Count, MaxHistory) ->
     Runnable0 = pending_jobs(),
     Runnable1 = lists:sort(fun oldest_job_first/2, Runnable0),
     Runnable2 = lists:filter(fun not_recently_crashed/1, Runnable1),
     Runnable3 = lists:sublist(Runnable2, Count),
-    lists:foreach(fun start_job_int/1, Runnable3).
+    lists:foreach(fun(J) -> start_job_int(J, MaxHistory) end, Runnable3).
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: that's a real shame, breaks hot upgrade. can't think of a way around it right now but :(

File: src/couch_replicator_scheduler.erl [hunk 19]
```diff
@@ -317,38 +334,38 @@ job_by_id(Id) ->
...
+    stop_jobs(Running - Max, MaxHistory);
 
-stop_excess_jobs(_, _) ->
+stop_excess_jobs(_, _, _) ->
     ok.
 
 
-start_pending_jobs(Max) ->
-    start_pending_jobs(Max, running_job_count(), pending_job_count()).
+start_pending_jobs(Max, MaxHistory) ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: 'Max' is now ambiguous, 'MaxJobs'?

File: src/couch_replicator_scheduler.erl [hunk 20]
```diff
@@ -317,38 +334,38 @@ job_by_id(Id) ->
...
+    start_jobs(Max - Running, MaxHistory);
 
-start_pending_jobs(_, _, _) ->
+start_pending_jobs(_, _, _, _) ->
     ok.
 
-rotate_jobs(MaxJobs, MaxChurn, Running, Pending) when Running == MaxJobs, Pending > 0 ->
-    stop_jobs(min([Pending, Running, MaxChurn])),
-    start_jobs(min([Pending, Running, MaxChurn]));
+rotate_jobs(MaxJobs, MaxChurn, Running, Pending, MaxHistory) when Running == MaxJobs, Pending > 0 ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: add it after MaxChurn, that's just untidy.

File: src/couch_replicator_scheduler_job.erl [hunk 21]
```diff
@@ -450,11 +450,8 @@ format_status(_Opt, [_PDict, State]) ->
     [{data, [{"State", state_strip_creds(State)}]}].
 
 
-report_job_error(_Rep, _Error) ->
-    % TODO: handle errors back to job scheduler to let it
-    % decide if needed to backoff and retry and what the back-off schedule
-    % should be
-    ok.
+report_job_error(Rep, Error) ->
+    exit({error, Rep, Error}).
```
Comment (rnewson) [inline src/couch_replicator_scheduler_job.erl]: this seems to have no bearing on the change, what's this for?

**Comment by rnewson** [comment]
as a meta comment, I think it might be time we pass State into most of the private functions. we're now passing three bits of it down, in addition to other parameters.

File: src/couch_replicator_scheduler.erl [hunk 22]
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
...
 
 not_recently_crashed(#job{} = Job) ->
-    case crash_history(Job) of
+    case Job#job.history of
         [] ->
             true;
-        [{crashed, When} | _] ->
-            timer:now_diff(os:timestamp(), When)
-                >= ?MINIMUM_CRASH_INTERVAL
+        [stopped | _] ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: `{stopped, _When}`, right?

File: src/couch_replicator_scheduler.erl [hunk 23]
```diff
@@ -205,52 +213,62 @@ handle_config_terminate(Self, _, _) ->
...
+            true;
+        _ ->
+            EventsBeforeStop = lists:takewhile(
+                fun({Event, _}) -> Event =/= stopped end, Job#job.history),
+            Crashes = lists:filter(
+                fun({{crashed, _Reason}, _When}) -> true; (_) -> false end,
+                EventsBeforeStop
+            ),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: why lists:filter instead of the simpler and shorter list comprehension?

File: src/couch_replicator_scheduler.erl (hunk 17 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Check latest `63012-scheduler` branch (rebase perhaps), there were changes  related to monitoring there.

File: src/couch_replicator_scheduler.erl (hunk 18 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: A list comprehension could do it then. Just return ok, in the end to stay consistent with the existing return type

File: src/couch_replicator_scheduler_job.erl (hunk 21 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler_job.erl]: It'll cause the job to terminate in a non-successful way, which is necessary for exponential backoff.

File: src/couch_replicator_scheduler.erl (hunk 23 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: The comprehension was getting a bit long, and I prefer multiline funs rather than multiline list comprehensions, but I could change it back to a long list comprehension.

File: src/couch_replicator_scheduler.erl (hunk 22 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Whoops

File: src/couch_replicator_scheduler.erl (hunk 18 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Yeah, a LC would work. I'll change that.

File: src/couch_replicator_scheduler_job.erl (hunk 21 above)
Comment (sagelywizard) [inline src/couch_replicator_scheduler_job.erl]: Though I'm a bit on the fence about passing the `Rep` state back, as it can be pretty big. The reason for the error is useful/necessary though.

**Comment by nickva** [comment]
There is a conflict with 63012-scheduler in the `handle_info({'DOWN', ...)` function clause

**Comment by nickva** [comment]
Unit tests pass on this branch

```
=======================================================
  All 171 tests passed.
```

Did manual testing. Exponential back-off seems to work (manually removed a target db) and so job staying in the crashed state for a while even after adding it back.

A few more comments above though, about `exit({error, ..})`

File: src/couch_replicator_scheduler_job.erl (hunk 21 above)
Comment (nickva) [inline src/couch_replicator_scheduler_job.erl]: Tested locally. Killed a replication by hand noticed that keeping `#rep{}` record in each crash event could be a problem. I leaks credentials to logs and is a rather large chunk of data (compared to other things there).

Also not sure if we should exit inside the context of the terminate/2 callback. It is a bit unusual. This was a place where we wrote errors back to documents. But if we don't plan on keeping #rep{} process will already exit with `Reason` so we can just not do anything. In a separate pr can remove report_job_error altogether.

File: src/couch_replicator_scheduler.erl [hunk 24]
```diff
@@ -317,38 +344,38 @@ job_by_id(Id) ->
...
+-spec reschedule(MaxJobs :: non_neg_integer(), MaxChurn :: non_neg_integer(), MaxHistory :: non_neg_integer()) -> ok.
+reschedule(MaxJobs, MaxChurn, MaxHistory)
   when is_integer(MaxJobs), MaxJobs >= 0, is_integer(MaxChurn), MaxChurn > 0 ->
     Running = running_job_count(),
     Pending = pending_job_count(),
-    stop_excess_jobs(MaxJobs, Running),
-    start_pending_jobs(MaxJobs, Running, Pending),
-    rotate_jobs(MaxJobs, MaxChurn, Running, Pending).
+    stop_excess_jobs(MaxJobs, Running, MaxHistory),
+    start_pending_jobs(MaxJobs, MaxHistory, Running, Pending),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: inconsistent parameter ordering.

File: src/couch_replicator_scheduler_job.erl (hunk 21 above)
Comment (rnewson) [inline src/couch_replicator_scheduler_job.erl]: it's unnecessary. killing a running replication job extends the history with a crash event without this change. remove report_job_error entirely, it does nothing.

**Comment by rnewson** [comment]
needs an extra clause in handle_config_change to cast to set_max_history

**Comment by sagelywizard** [comment]
Yet another rebase.

File: src/couch_replicator_scheduler.erl [hunk 25]
```diff
@@ -320,40 +345,44 @@ job_by_id(Id) ->
...
+    #state{max_jobs=MaxJobs, max_history=MaxHistory} = State,
+    if Running < MaxJobs, Pending > 0 ->
+        start_jobs(MaxJobs - Running, MaxHistory);
+    true ->
+        ok
+    end.
 
+rotate_jobs(State, Running, Pending) ->
+    #state{max_jobs=MaxJobs, max_history=MaxHistory, max_churn=MaxChurn} = State,
+    if Running == MaxJobs, Pending > 0 ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:380]: extract MaxJobs in the function head so the guard clause can work, and extract the others in that `= State` line.

**Comment by rnewson** [comment]
+1

**Comment by sagelywizard** [comment]
That was the squash/rebase.
